git-branchless/src/commands/smartlog.rs

```diff
@@ -151,41 +171,128 @@ mod graph {
-                graph.iter().filter_map(
-                    |(child_oid, node)| if !node.is_main { Some(child_oid) } else { None },
-                );
-
-            let mut links = Vec::new();
-            for child_oid in non_main_node_oids {
-                let parent_vertexes = dag.query().parents(CommitSet::from(*child_oid))?;
-                let parent_oids = commit_set_to_vec(&parent_vertexes)?;
-                for parent_oid in parent_oids {
-                    if graph.contains_key(&parent_oid) {
-                        links.push((*child_oid, parent_oid))
+        let mut immediate_links: Vec<(NonZeroOid, NonZeroOid)> = Vec::new();
+        let mut non_immediate_links: Vec<(NonZeroOid, NonZeroOid)> = Vec::new();
+
+        let non_main_node_oids =
+            graph
+                .iter()
+                .filter_map(|(child_oid, node)| if !node.is_main { Some(child_oid) } else { None });
+
+        let graph_vertices: CommitSet = graph.keys().cloned().collect();
+        for child_oid in non_main_node_oids {
+            let parent_vertices = dag.query().parents(CommitSet::from(*child_oid))?;
+
+            // Find immediate parent-child links.
+            let (parent_oids, parents_in_graph) = match smartlog_variant {
+                SmartlogVariant::Connected => {
+                    let parent_oids = commit_set_to_vec(&parent_vertices)?;
+                    (parent_oids, CommitSet::empty())
+                }
+                SmartlogVariant::Sparse => {
+                    let parents_in_graph = parent_vertices.intersection(&graph_vertices);
+                    let parent_oids = commit_set_to_vec(&parents_in_graph)?;
+                    (parent_oids, parents_in_graph)
+                }
+            };
+            for parent_oid in parent_oids {
+                immediate_links.push((*child_oid, parent_oid))
+            }
+            match smartlog_variant {
+                SmartlogVariant::Connected => {
+                    // No need to find non-immediate links.
+                }
+                SmartlogVariant::Sparse
+                    if parent_vertices.count()? == parents_in_graph.count()? =>
+                {
+                    // All parents have been handled, no need to look for
+                    // non-immediate links
+                }
+                SmartlogVariant::Sparse => {
+                    // Find non-immediate ancestor links.
+                    let excluded_parents = parent_vertices.difference(&graph_vertices);
+                    let excluded_parent_oids = commit_set_to_vec(&excluded_parents)?;
+                    for parent_oid in excluded_parent_oids {
+                        // Find the nearest ancestor that is included in the graph and
+                        // also on the same branch.
+
+                        let nearest_branch_ancestor = match dag
+                            .find_path_to_main_branch(effects, CommitSet::from(parent_oid))?
```

Comment:
If possible, we should delete the `find_path_to_main_branch` function and use `dag.query().range()` instead. I wrote this function before we started using the segmented changelog DAG, and calculating the real range of commits was a pain, so I just wrote code to calculate a single path. (I think there's also a function about calculating merge-bases that can be deleted in favor of `gca_one()` and `gca_all()`.)

Reply:
OK, done. I removed both `find_path_to_main_branch()` and `get_one_merge_base_oid()`. That's in a separate commit. Note though that I did not get into `gca_one` vs `gca_all` because for the purposes in this PR I think we don't actually care about multiple bases; just the nearest ancestor between a merge base and the commit. Also, given the difficulty in rendering merge commits, I'm not even sure what we would render if a commit has 2 disconnected ancestors in a sparse log.

If you'd like me to look into that more, I can; it's just not a case that I'm familiar with.